client/wsclient.go

```diff
@@ -413,3 +426,46 @@ func (c *wsClient) runUntilStopped(ctx context.Context) {
 		c.runOneCycle(ctx)
 	}
 }
+
+// useProxy sets the websocket dialer to use the passed proxy URL.
+// If the proxy has no schema http is used.
+// This method is not thread safe and must be called before c.dialer is used.
+func (c *wsClient) useProxy(proxy string, headers http.Header, cfg *tls.Config) error {
+	proxyURL, err := url.Parse(proxy)
+	if err != nil || proxyURL.Scheme == "" || proxyURL.Host == "" { // error or bad URL - try to use http as scheme to resolve
+		proxyURL, err = url.Parse("http://" + proxy)
+		if err != nil {
+			return err
+		}
+	}
+	if proxyURL.Hostname() == "" {
+		return url.InvalidHostError(proxy)
+	}
+
+	// Clear previous settings
+	c.dialer.Proxy = nil
+	c.dialer.NetDialContext = nil
+	c.dialer.NetDialTLSContext = nil
+
+	switch strings.ToLower(proxyURL.Scheme) {
+	case "http":
+		// FIXME: dialer.NetDialContext is currently used as a work around instead of setting dialer.Proxy as gorilla/websockets does not have 1st class support for setting proxy connect headers
+		// Once http://github.com/gorilla/websocket/issues/479 is complete, we should use dialer.Proxy, and dialer.ProxyConnectHeader
+		dialer, err := dialer.NewProxyConnectDialer(proxyURL, &net.Dialer{}, dialer.WithProxyConnectHeaders(headers))
+		if err != nil {
+			return err
+		}
+		c.dialer.NetDialContext = dialer.DialContext
+	case "https":
+		dialer, err := dialer.NewProxyConnectDialer(proxyURL, &net.Dialer{}, dialer.WithTLS(cfg), dialer.WithProxyConnectHeaders(headers))
+		if err != nil {
+			return err
+		}
+		c.dialer.NetDialTLSContext = dialer.DialContext
```

Comment:
This is a little messy (as we configure the `websockets.Dialer` to behave slightly differently for HTTP(S) proxies vs non-HTTP(S) proxies; but it's much easer to do at this point in time then it is to replace the entire websockets library.

If https://github.com/gorilla/websocket/pull/988 is ever merged and released we can clean this up a lot.